Source/ui_unix/GamePad/GamePadDeviceListener.cpp

```diff
@@ -0,0 +1,230 @@
+	UpdateDeviceList();
+
+}
+
+CGamePadDeviceListener::CGamePadDeviceListener(bool filter)
+	: m_filter(filter)
+{
+	m_thread = std::thread(std::bind(&CGamePadDeviceListener::InputDeviceListenerThread, this));
+	UpdateDeviceList();
+
+}
+
+CGamePadDeviceListener::~CGamePadDeviceListener()
+{
+	m_running = false;
+	m_GPIEList.clear();
+	if(m_thread.joinable()) m_thread.join();
+}
+
+void CGamePadDeviceListener::RePopulateAbs()
+{
+	for (auto& GPI : m_GPIEList)
+		GPI.second.get()->RePopulateAbs();
+}
+
+void CGamePadDeviceListener::UpdateOnInputEventCallback(CGamePadDeviceListener::OnInputEvent OnInputEventFunction)
+{
+	OnInputEventCallBack = OnInputEventFunction;
+	for (auto& device : m_devicelist)
+	{
+		for (auto& GPI : m_GPIEList)
+		{
+			if(GPI.first == device.first)
+			{
+				GPI.second.get()->OnInputEvent.disconnect_all_slots();
+				GPI.second.get()->OnInputEvent.connect(OnInputEventCallBack);
+				goto nextdev;
+			}
+		}
+		{
+			auto GamePadInput = std::make_unique<CGamePadInputEventListener>(device.second.path, m_filter);
+			GamePadInput->OnInputEvent.connect(OnInputEventCallBack);
+			m_GPIEList.emplace(device.first, std::move(GamePadInput));
+		}
+nextdev:
+		continue;
+	}
+}
+
+void CGamePadDeviceListener::DisconnectInputEventCallback()
+{
+	OnInputEventCallBack = nullptr;
+	for (auto& GPI : m_GPIEList)
+	{
+		GPI.second.get()->OnInputEvent.disconnect_all_slots();
+	}
+}
+
+void CGamePadDeviceListener::UpdateDeviceList()
+{
+	std::string path = "/dev/input/";
+	for (auto & p : fs::directory_iterator(path))
+		if (p.path().filename().string().find("event") != std::string::npos)
+		{
+			AddDevice(p.path());
+		}
+}
+
+void CGamePadDeviceListener::AddDevice(const fs::path& path)
+{
+	inputdev_pair idp;
+	if(IsValidDevice(path, idp))
+	{
+		m_devicelist.insert(idp);
+		if (OnInputEventCallBack)
+		{
+			auto GamePadInput = std::make_unique<CGamePadInputEventListener>(idp.second.path, m_filter);
+			GamePadInput->OnInputEvent.connect(OnInputEventCallBack);
+			m_GPIEList.emplace(idp.first, std::move(GamePadInput));
+		}
+	}
+}
+
+void CGamePadDeviceListener::RemoveDevice(char* key)
+{
+	m_devicelist.erase(key);
+	m_GPIEList.erase(key);
+}
+
+void CGamePadDeviceListener::InputDeviceListenerThread()
+{
+	int fd = inotify_init1(IN_NONBLOCK);
+	if ( fd < 0 ) {
+		perror("inotify_init");
+	}
+
+	struct pollfd fds[2];
+	sigset_t mask;
+
+	fds[0].fd = fd;
+	fds[0].events = POLLIN;
+
+	sigemptyset(&mask);
+	sigaddset(&mask, SIGINT);
+	fds[1].fd = signalfd(-1, &mask, SFD_NONBLOCK);
+	fds[1].events = POLLIN;
+
+	sigprocmask(SIG_BLOCK, &mask, NULL);
+
+	std::string devinput_dir("/dev/input");
+	int wd = inotify_add_watch(fd, devinput_dir.c_str(), WATCH_FLAGS);
+	char buffer[EVENT_BUF_LEN];
+
+	while (m_running)
+	{
+		do
+		{
+			if (!m_running) goto cleanup;
+		} while(poll(fds, 2, 500) == 0);
+
+		int length = read(fd, buffer, EVENT_BUF_LEN);
+
+		if ( length < 0 )
+		{
+			perror("read");
```

Comment:
not through the remaining of the loop, but calling `continue` shouldn't be an issue.